File: utils/risk_scoring.py
"""Transparent risk scoring and red-flag detection."""

from __future__ import annotations

import re
from difflib import SequenceMatcher

import pandas as pd

from utils.constants import (
    CONSULTANCY_KEYWORDS,
    CONSULTANCY_SUPPLIERS,
    LEGAL_KEYWORDS,
    RED_FLAG_DEFINITIONS,
)


def _normalize_text(text: str) -> str:
    text = str(text).lower()
    text = re.sub(r"[^a-z0-9\s]", " ", text)
    return re.sub(r"\s+", " ", text).strip()


def _similarity(a: str, b: str) -> float:
    return SequenceMatcher(None, _normalize_text(a), _normalize_text(b)).ratio()


def _is_consultancy(row: pd.Series) -> bool:
    cpv = str(row.get("cpv_code", ""))
    title = str(row.get("title", "")).lower()
    desc = str(row.get("description", "")).lower()
    supplier = str(row.get("supplier", ""))
    if cpv.startswith("79") and not cpv.startswith("791"):
        return True
    if supplier in CONSULTANCY_SUPPLIERS:
        return True
    return any(k in title or k in desc for k in CONSULTANCY_KEYWORDS)


def _is_legal(row: pd.Series) -> bool:
    cpv = str(row.get("cpv_code", ""))
    title = str(row.get("title", "")).lower()
    desc = str(row.get("description", "")).lower()
    if cpv.startswith("791"):
        return True
    return any(k in title or k in desc for k in LEGAL_KEYWORDS)


def detect_duplicate_risk(df: pd.DataFrame) -> pd.Series:
    flags = []
    for idx, row in df.iterrows():
        award = row["award_date"]
        window_start = award - pd.Timedelta(days=730)
        peers = df[
            (df.index != idx)
            & (df["supplier"] == row["supplier"])
            & (df["buyer"] == row["buyer"])
            & (df["award_date"] >= window_start)
            & (df["award_date"] <= award)
        ]
        hit = False
        for _, peer in peers.iterrows():
            title_sim = _similarity(row["title"], peer["title"])
            desc_sim = _similarity(row["description"], peer["description"])
            if title_sim >= 0.88 or (title_sim >= 0.75 and desc_sim >= 0.82):
                hit = True
                break
        flags.append(hit)
    return pd.Series(flags, index=df.index)


def detect_repeated_consultancy(df: pd.DataFrame) -> pd.Series:
    flags = []
    for idx, row in df.iterrows():
        if not _is_consultancy(row):
            flags.append(False)
            continue
        award = row["award_date"]
        window_start = award - pd.Timedelta(days=548)
        peers = df[
            (df.index != idx)
            & (df["buyer"] == row["buyer"])
            & (df["award_date"] >= window_start)
            & (df["award_date"] <= award)
        ]
        consult_count = sum(1 for _, p in peers.iterrows() if _is_consultancy(p))
        same_cat = sum(
            1
            for _, p in peers.iterrows()
            if _is_consultancy(p) and p["category"] == row["category"]
        )
        flags.append(consult_count >= 4 or same_cat >= 3)
    return pd.Series(flags, index=df.index)


def detect_value_spike(df: pd.DataFrame) -> pd.Series:
    flags = []
    df = df.copy()
    df["award_year"] = df["award_date"].dt.year
    for idx, row in df.iterrows():
        recent = df[
            (df.index != idx)
            & (df["buyer"] == row["buyer"])
            & (df["category"] == row["category"])
            & (df["award_year"] >= row["award_year"] - 3)
            & (df["award_year"] <= row["award_year"])
        ]
        if len(recent) < 8:
            flags.append(False)
            continue
        median_val = recent["value_gbp"].median()
        if median_val <= 0:
            flags.append(False)
            continue
        flags.append(row["value_gbp"] > median_val * 4 and row["value_gbp"] > 250_000)
    return pd.Series(flags, index=df.index)


def compute_risk_scores(df: pd.DataFrame) -> pd.DataFrame:
    """Add red flag columns and composite risk score to dataframe."""
    result = df.copy()
    result["award_date"] = pd.to_datetime(result["award_date"])

    dup = detect_duplicate_risk(result)
    consult = detect_repeated_consultancy(result)
    spike = detect_value_spike(result)

    result["flag_duplicate_risk"] = dup
    result["flag_repeated_consultancy"] = consult
    result["flag_value_spike"] = spike

    flag_cols = ["flag_duplicate_risk", "flag_repeated_consultancy", "flag_value_spike"]
    result["red_flag_count"] = result[flag_cols].sum(axis=1).astype(int)
    result["red_flags"] = result.apply(_format_red_flags, axis=1)

    weights = [RED_FLAG_DEFINITIONS["duplicate_risk"]["weight"],
               RED_FLAG_DEFINITIONS["repeated_consultancy"]["weight"],
               RED_FLAG_DEFINITIONS["value_spike"]["weight"]]
    result["risk_score"] = (
        result["flag_duplicate_risk"].astype(int) * weights[0]
        + result["flag_repeated_consultancy"].astype(int) * weights[1]
        + result["flag_value_spike"].astype(int) * weights[2]
    )

    # Severity bump for very high values
    p95 = result["value_gbp"].quantile(0.95)
    result.loc[result["value_gbp"] > p95, "risk_score"] = (
        result.loc[result["value_gbp"] > p95, "risk_score"] + 10
    ).clip(0, 100)

    result["is_legal"] = result.apply(_is_legal, axis=1)
    result["is_consultancy"] = result.apply(_is_consultancy, axis=1)
    result["risk_score"] = result["risk_score"].clip(0, 100).astype(int)
    return result


def _format_red_flags(row: pd.Series) -> str:
    active = []
    if row["flag_duplicate_risk"]:
        active.append("Duplicate")
    if row["flag_repeated_consultancy"]:
        active.append("Consultancy")
    if row["flag_value_spike"]:
        active.append("Value Spike")
    return ", ".join(active) if active else "—"


def explain_flags(row: pd.Series) -> str:
    parts = []
    if row["flag_duplicate_risk"]:
        parts.append(RED_FLAG_DEFINITIONS["duplicate_risk"]["description"])
    if row["flag_repeated_consultancy"]:
        parts.append(RED_FLAG_DEFINITIONS["repeated_consultancy"]["description"])
    if row["flag_value_spike"]:
        parts.append(RED_FLAG_DEFINITIONS["value_spike"]["description"])
    return " ".join(parts) if parts else "No red flags detected for this contract."

File: lib/risk-scoring.ts
import {
  CONSULTANCY_KEYWORDS,
  CONSULTANCY_SUPPLIERS,
  DEPARTMENT_TAGS,
  LEGAL_KEYWORDS,
  RED_FLAG_DEFINITIONS,
} from "./constants";
import type { Contract, RawContract } from "./types";

function normalizeText(text: string): string {
  return text
    .toLowerCase()
    .replace(/[^a-z0-9\s]/g, " ")
    .replace(/\s+/g, " ")
    .trim();
}

function similarity(a: string, b: string): number {
  const s1 = normalizeText(a);
  const s2 = normalizeText(b);
  if (s1 === s2) return 1;
  if (!s1 || !s2) return 0;

  const len1 = s1.length;
  const len2 = s2.length;
  const matrix: number[][] = Array.from({ length: len1 + 1 }, (_, i) =>
    Array.from({ length: len2 + 1 }, (_, j) => (i === 0 ? j : j === 0 ? i : 0))
  );

  for (let i = 1; i <= len1; i++) {
    for (let j = 1; j <= len2; j++) {
      const cost = s1[i - 1] === s2[j - 1] ? 0 : 1;
      matrix[i][j] = Math.min(
        matrix[i - 1][j] + 1,
        matrix[i][j - 1] + 1,
        matrix[i - 1][j - 1] + cost
      );
    }
  }

  const dist = matrix[len1][len2];
  return 1 - dist / Math.max(len1, len2);
}

function isConsultancy(row: RawContract): boolean {
  const cpv = String(row.cpv_code ?? "");
  const title = row.title.toLowerCase();
  const desc = row.description.toLowerCase();
  if (cpv.startsWith("79") && !cpv.startsWith("791")) return true;
  if (CONSULTANCY_SUPPLIERS.includes(row.supplier)) return true;
  return CONSULTANCY_KEYWORDS.some((k) => title.includes(k) || desc.includes(k));
}

function isLegal(row: RawContract): boolean {
  const cpv = String(row.cpv_code ?? "");
  const title = row.title.toLowerCase();
  const desc = row.description.toLowerCase();
  if (cpv.startsWith("791")) return true;
  return LEGAL_KEYWORDS.some((k) => title.includes(k) || desc.includes(k));
}

function daysBetween(a: string, b: string): number {
  return Math.abs(new Date(a).getTime() - new Date(b).getTime()) / 86400000;
}

function detectDuplicateRisk(contracts: RawContract[]): boolean[] {
  return contracts.map((row, idx) => {
    const peers = contracts.filter(
      (p, i) =>
        i !== idx &&
        p.supplier === row.supplier &&
        p.buyer === row.buyer &&
        daysBetween(p.award_date, row.award_date) <= 730 &&
        new Date(p.award_date) <= new Date(row.award_date)
    );
    return peers.some((peer) => {
      const titleSim = similarity(row.title, peer.title);
      const descSim = similarity(row.description, peer.description);
      return titleSim >= 0.88 || (titleSim >= 0.75 && descSim >= 0.82);
    });
  });
}

function detectRepeatedConsultancy(contracts: RawContract[]): boolean[] {
  return contracts.map((row, idx) => {
    if (!isConsultancy(row)) return false;
    const peers = contracts.filter(
      (p, i) =>
        i !== idx &&
        p.buyer === row.buyer &&
        daysBetween(p.award_date, row.award_date) <= 548 &&
        new Date(p.award_date) <= new Date(row.award_date)
    );
    const consultCount = peers.filter(isConsultancy).length;
    const sameCat = peers.filter(
      (p) => isConsultancy(p) && p.category === row.category
    ).length;
    return consultCount >= 4 || sameCat >= 3;
  });
}

function detectValueSpike(contracts: RawContract[]): boolean[] {
  return contracts.map((row, idx) => {
    const year = new Date(row.award_date).getFullYear();
    const recent = contracts.filter(
      (p, i) =>
        i !== idx &&
        p.buyer === row.buyer &&
        p.category === row.category &&
        new Date(p.award_date).getFullYear() >= year - 3 &&
        new Date(p.award_date).getFullYear() <= year
    );
    if (recent.length < 8) return false;
    const values = recent.map((p) => p.value_gbp).sort((a, b) => a - b);
    const mid = Math.floor(values.length / 2);
    const median =
      values.length % 2 === 0
        ? (values[mid - 1] + values[mid]) / 2
        : values[mid];
    if (median <= 0) return false;
    return row.value_gbp > median * 4 && row.value_gbp > 250_000;
  });
}

function formatRedFlags(flags: {
  duplicate: boolean;
  consultancy: boolean;
  spike: boolean;
}): string {
  const active: string[] = [];
  if (flags.duplicate) active.push("Duplicate");
  if (flags.consultancy) active.push("Consultancy");
  if (flags.spike) active.push("Value Spike");
  return active.length ? active.join(", ") : "—";
}

export function explainFlags(contract: Contract): string {
  const parts: string[] = [];
  if (contract.flag_duplicate_risk)
    parts.push(RED_FLAG_DEFINITIONS.duplicate_risk.description);
  if (contract.flag_repeated_consultancy)
    parts.push(RED_FLAG_DEFINITIONS.repeated_consultancy.description);
  if (contract.flag_value_spike)
    parts.push(RED_FLAG_DEFINITIONS.value_spike.description);
  return parts.length
    ? parts.join(" ")
    : "No red flags detected for this contract.";
}

export function computeRiskScores(raw: RawContract[]): Contract[] {
  const dup = detectDuplicateRisk(raw);
  const consult = detectRepeatedConsultancy(raw);
  const spike = detectValueSpike(raw);

  const values = raw.map((r) => r.value_gbp).sort((a, b) => a - b);
  const p95Idx = Math.floor(values.length * 0.95);
  const p95 = values[p95Idx] ?? 0;

  return raw.map((row, i) => {
    let riskScore =
      (dup[i] ? RED_FLAG_DEFINITIONS.duplicate_risk.weight : 0) +
      (consult[i] ? RED_FLAG_DEFINITIONS.repeated_consultancy.weight : 0) +
      (spike[i] ? RED_FLAG_DEFINITIONS.value_spike.weight : 0);

    if (row.value_gbp > p95) riskScore = Math.min(100, riskScore + 10);

    const redFlagCount =
      Number(dup[i]) + Number(consult[i]) + Number(spike[i]);

    return {
      ...row,
      department_tag: row.department_tag ?? DEPARTMENT_TAGS[row.buyer] ?? "Other",
      contracts_finder_url:
        row.contracts_finder_url ??
        `https://www.contractsfinder.service.gov.uk/Notice/${row.notice_id}`,
      flag_duplicate_risk: dup[i],
      flag_repeated_consultancy: consult[i],
      flag_value_spike: spike[i],
      red_flag_count: redFlagCount,
      red_flags: formatRedFlags({
        duplicate: dup[i],
        consultancy: consult[i],
        spike: spike[i],
      }),
      risk_score: Math.min(100, riskScore),
      is_legal: isLegal(row),
      is_consultancy: isConsultancy(row),
    };
  });
}